### PR TITLE
Go back to building grpc-haskell-core using c2hs

### PR DESCRIPTION
**Pull request by cocreature**
This should hopefully avoid issues like the CSize vs CULong issue we
had a while back and might fix some of the issues we have been seeing
on CI.

I’m marking the Haskell ledger bindings as non-flaky for now so we can
see if the issues reappear.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: bazel_tools/haskell-c2hs-windows.patch [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+             set.to_depset(version_macro_headers),
+         ]),
+         outputs = [hs_file, chi_file],
+-        command = """
++        command =
++        # cpp (called via c2hs) gets very unhappy if the mingw bin dir is
++        # not in PATH so we add it to PATH explicitely.
++        (
++        """
++        export PATH=$PATH:{cc}/..
```
Comment (aherrmann-da) [inline bazel_tools/haskell-c2hs-windows.patch]: Is that a legal path? `{cc}` points to a file. If I try
```
touch foo
ls foo/..
```
I get "not a directory".

Comment (cocreature) [reply bazel_tools/haskell-c2hs-windows.patch]: That is a very good point, confusingly it seems to work …

File: bazel_tools/haskell-c2hs-windows.patch [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+     args.add_all(["-C" + x for x in cc.include_args])
++    args.add_all(ctx.attr.extra_args)
+ 
+     dep_chi_files = [
+         dep[C2hsLibraryInfo].chi_file
+@@ -59,22 +61,36 @@ def _c2hs_library_impl(ctx):
+         (version_macro_headers, version_macro_flags) = version_macro_includes(dep_info)
+         args.add_all(["-C" + x for x in version_macro_flags])
+ 
++    (inputs, input_manifests) = ctx.resolve_tools(tools = [c2hs])
```
Comment (cocreature) [inline bazel_tools/haskell-c2hs-windows.patch]: @aherrmann-da I need `c2hs` to be a target for this to work but I couldn’t figure out how I could then get the path from the target so as a hacky workaround I added `c2hs_exe` which points to the file and `c2hs` which points to the target to the toolchain. What’s the non-hacky way to accomplish this?

File: bazel_tools/haskell-c2hs-windows.patch (hunk 1 above)
Comment (cocreature) [reply bazel_tools/haskell-c2hs-windows.patch]: Switched to `paths.dirname(cc)` which seems less stupid.

File: bazel_tools/haskell-c2hs-windows.patch (hunk 2 above)
Comment (aherrmann-da) [reply bazel_tools/haskell-c2hs-windows.patch]: Yes, the information that it's a single executable file is unfortunately lost once you only have the `Target`. The separate `c2hs_exe` seems like a good solution. I've essentially done the same [with the cc_wrapper](https://github.com/tweag/rules_haskell/pull/1039/files#diff-bd788ee16d6ba21ea60c42efea01aba3R163-R166).

File: nix/third-party/gRPC-haskell/core/BUILD.bazel [hunk 3]
```diff
@@ -10,11 +10,22 @@ load("//bazel_tools:fat_cc_library.bzl", "fat_cc_library")
 # FIXME: The files in the chs directory should be generated on-the-fly using
```
Comment (aherrmann-da) [inline nix/third-party/gRPC-haskell/core/BUILD.bazel]: The comment can be removed now.

Comment (cocreature) [reply nix/third-party/gRPC-haskell/core/BUILD.bazel]: Good catch, fixed!

**Comment by cocreature** [comment]
Marked the Haskell ledger bindings flaky again since I’m able to make them fail locally but at least I haven’t seen the gRPC issues that we’ve seen before.

**Pull request merged**